processor/src/lib.rs

```diff
@@ -152,44 +152,44 @@ where
 // ================================================================================================
 
 #[cfg(not(any(test, feature = "internals")))]
-struct Process<A>
+struct Process<H>
 where
-    A: AdviceProvider,
+    H: Host,
 {
     system: System,
     decoder: Decoder,
     stack: Stack,
     range: RangeChecker,
     chiplets: Chiplets,
-    advice_provider: A,
+    advice_provider: RefCell<H>,
     max_cycles: u32,
 }
```

Comment:
good catch I will rename.  

We use a `RefCell` as we have an immutable and mutable reference to fields of the `Process` object at the same time due to: 

```rust
pub trait Host {
    fn execute_host_function<S: ProcessState>(
        &mut self,
        process: &S,
        function: &HostFunction,
    ) -> Result<usize, ExecutionError>;
```

The rust compiler cant seem to assert that the immutable and mutable borrows point to different regions of memory and as such we need to defer the borrow checks to runtime using a `RefCell` with interior mutability.  This does come withe some runtime overhead. See the following invocation, we borrow `host` as mutable and then we borrow `self` (`self as ProcessState`) as immutable:

https://github.com/0xPolygonMiden/miden-vm/blob/a5ff893a25693a3a76390dba3b4bd68f04a7f1eb/processor/src/operations/crypto_ops.rs#L79-L82

Reply:
I've implemented a solution that does not use `RefCell`.  Instead we introduce a struct that allows the compiler to satisfy compile time borrow checker requirements. https://github.com/0xPolygonMiden/miden-vm/pull/1089

Reply:
Ah - I see! The solution in #1089 works, but it does make interactions with the host even more cumbersome, and I'm not sure if runtime overhead of `RefCell` is significant enough for us to make this trade off.

Another option is to separate VM state fields in the `Process` from the host field - e.g., something like this:
```Rust
struct Process<H>
where
    H: Host,
{
    state: VmState,
    host: H,
    max_cycles: u32,
}
```
But this would cause a huge number of changes - so, probably not a viable solution.

Reply:
Yeah I think the change you proposed is preferable but probably not practical without investing a significant amount of time.  How about we leave the `RefCell` for now and create an issue to consider refactoring to your solution in the future?

Reply:
Yes - agreed!